A36-xds-for-servers.md

```diff
@@ -0,0 +1,324 @@
+continue after the listener is closed and Serve() returns.
+
+Because service registration is done on a method in the generated code, and
+`grpc.Server` is a struct, the underlying `grpc.Server` will need to be exposed.
+
+
+```
+// instead of
+s := grpc.NewServer()
+pb.RegisterGreeterServer(s, &server{})
+// it'd be
+s := xds.NewGRPCServer()
+pb.RegisterGreeterServer(s.Server(), &server{})
+```
+
+
+To allow passing configuration, the `xds.Server` will be responsible for
+creating the actual `XdsTransportCredentials`. There will be a façade
+`XdsTransportCredentials` for the user to provide as a `ServerOption`.
+`xds.Server` will look through the `ServerOptions`, do a type assertion on the
+`TransportCredential`, and if it is the façade `XdsTransportCredentials` it will
+replace the credentials with the real `XdsTransportCredential`. The real
+`XdsTransportCredential` will need to use the local IP to find the port-specific
+configuration.
+
+
+## Rationale
+
+XdsServer server startup is delayed waiting on initial configuration from xDS to
+avoid applications getting hung in the rare event of an issue during startup. If
+the start API was one-shot, then users would need to loop themselves and many
+might introduce bugs in the rarely-run code path or simply not handle the case
+at all.
+
+Ideally `listen()` would be delayed until just before `accept()` to avoid
+hanging new client connections. However, the kernel does not fully initialize
+the port details until `listen()` and some gRPC implementations have behavior
+and APIs that makes this infeasible. This may cause serious problems for
+specific clients using pick-first but is essentially unavoidable today. The
+impact is hoped to be low as servers will generally receive their initial
+configuration quickly (on the order of a second) and most clients will likely be
+using a variant of round-robin. We'd also encourge the xDS control plane to
+avoid handing out addresses of not-yet-started servers. There are a few possible
+remedies, for example using [Happy Eyeballs][RFC8305] for pick-first or having
+the server `accept()` and immediately `close()` connections, but any such
+mitigations would be future work.
+
+[RFC8305]: https://tools.ietf.org/html/rfc8305
+
+We chose the XdsServer approach over an "xDS Interceptor." In such a design, the
+user would just construct an XdsInterceptor and add it to their server.
+Unfortunately the interceptor would not be informed of the server lifecycle and
+so could not manage the connection to the xDS control plane. It also is not
+informed of listening ports until RPCs actually arrive on those ports and is not
+able to learn when a port is no longer in use (as is possible in grpc-go). It
+would also be difficult to share the XdsClient configuration between the
+interceptor and credentials. It simply cannot be used in its current state.
+
+We chose the XdsServer approach over creating a new server-side plugin API. In
```

Comment:
Even with the `XdsServerBuilder` approach, we can offer a convenience function for ease of use. For example in  Java:
```
ServerBuilder<?> serverBuilder = Grpc.newServerBuilderForPort(host, port, serverCreds);
```
host is the new argument and if you pass something like `"xds:///localhost"` then the XdsServerBuilder is returned. Similar to how you engage xDS on the client side. Of course you will then have to cast the ServerBuilder to XdsServerBuilder if you want to use other functionality of the builder.